providers/common/ciphers/ciphers_locl.h

```diff
@@ -56,54 +97,60 @@ typedef struct prov_aes_key_st {
 #endif /* defined(OPENSSL_CPUID_OBJ) && defined(__s390__) */
     } plat;
 
-    /* The cipher functions we are going to use */
-    const PROV_AES_CIPHER *ciph;
-
-    /* The mode that we are using */
-    int mode;
-
-    /* Set to 1 if we are encrypting or 0 otherwise */
-    int enc;
-
-    unsigned char iv[AES_BLOCK_SIZE];
-
-    /*
-     * num contains the number of bytes of |iv| which are valid for modes that
-     * manage partial blocks themselves.
-     */
-    size_t num;
-
-    /* Buffer of partial blocks processed via update calls */
-    unsigned char buf[AES_BLOCK_SIZE];
-
-    /* Number of bytes in buf */
-    size_t bufsz;
-
-    uint64_t flags;
+} PROV_AES_KEY;
 
-    size_t keylen;
+typedef struct prov_aria_key_st {
+    PROV_GENERIC_KEY base;      /* Must be first */
+    union {
+        OSSL_UNION_ALIGN;
+        ARIA_KEY ks;
+    } ks;
+} PROV_ARIA_KEY;
 
-    /* Whether padding should be used or not */
-    unsigned int pad : 1;
-} PROV_AES_KEY;
+typedef struct prov_camellia_key_st {
+    PROV_GENERIC_KEY base;      /* Must be first */
+    union {
+        OSSL_UNION_ALIGN;
+        CAMELLIA_KEY ks;
```

Comment:
Is it worth merging these?  The difference is the key and it's already in a union -- include all keys in the union and use the unions address to reference them?

Reply:
Not sure - the aes one has a lot more platform specific stuff in it.. That would be the main difference..

Reply:
So I decided not to do this, because we would end up with a lot of stuff included in each struct (the AES can include a lot of S390x code) which would then be carried around in every cipher (on a s390x machine at least).